File: archive/go/2021/may/convert-sorted-list-to-binary-search-tree.go
package may

import "leetcode/common"

// TreeNode define
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 依次把链表的中间点作为根结点，类似二分的思想，递归排列所有结点
func sortedListToBST(head *common.ListNode) *TreeNode {
	if head == nil {
		return nil
	}
	if head != nil && head.Next == nil {
		return &TreeNode{Val: head.Val, Left: nil, Right: nil}
	}
	middleNode, preNode := middleNodeAndPreNode(head)
	if middleNode == nil {
		return nil
	}
	if preNode != nil {
		preNode.Next = nil
	}
	if middleNode == head {
		head = nil
	}
	return &TreeNode{Val: middleNode.Val, Left: sortedListToBST(head), Right: sortedListToBST(middleNode.Next)}
}

func middleNodeAndPreNode(head *common.ListNode) (middle *common.ListNode, pre *common.ListNode) {
	if head == nil || head.Next == nil {
		return nil, head
	}
	p1 := head
	p2 := head
	for p2.Next != nil && p2.Next.Next != nil {
		pre = p1
		p1 = p1.Next
		p2 = p2.Next.Next
	}
	return p1, pre
}

File: src/tree/convert_sorted_list_to_binary_search_tree.rs
use crate::{ListNode, Solution, TreeNode};

use std::cell::RefCell;
use std::rc::Rc;

type Node = Option<Rc<RefCell<TreeNode>>>;

impl Solution {
    pub fn sorted_list_to_bst(mut head: Option<Box<ListNode>>) -> Node {
        //  turn linked list into array -> vec into binary tree
        let mut stack = Vec::new();
        while let Some(node) = head {
            stack.push(node.val);
            head = node.next;
        }
        Solution::_sorted_array_to_bst(&stack[..])
    }
    fn _sorted_array_to_bst(nums: &[i32]) -> Node {
        let n = nums.len();

        match n {
            0 => None,
            _ => {
                let m = n / 2;
                let mut node = TreeNode::new(nums[m]);
                node.left = Solution::_sorted_array_to_bst(&nums[..m]);
                node.right = Solution::_sorted_array_to_bst(&nums[m + 1..]);

                Some(Rc::new(RefCell::new(node)))
            }
        }
    }
}
